packages/jsapi-utils/src/ViewportDataUtils.ts

```diff
@@ -0,0 +1,158 @@
+ * @param viewportData State object for managing a list of KeyedItem data.
+ * @param deserializeRow Converts a DH Row to an item object.
+ * @returns Handler function for a `dh.Table.EVENT_UPDATED` event.
+ */
+export function createOnTableUpdatedHandler<T>(
+  table: Table | TreeTable | null,
+  viewportData: ListData<KeyedItem<T>>,
+  deserializeRow: RowDeserializer<T>
+): (event: OnTableUpdatedEvent) => void {
+  /**
+   * Handler for a `dh.Table.EVENT_UPDATED` event.
+   */
+  return function onTableUpdated(event: OnTableUpdatedEvent) {
+    if (table == null) {
+      return;
+    }
+
+    const { offset, rows } = event.detail;
+
+    log.debug('table updated', event.detail);
+
+    rows.forEach(row => {
+      const item = deserializeRow(row, table.columns);
+
+      const keyedItem = {
+        key: createKeyFromOffsetRow(row, offset),
+        item,
+      };
+
+      if (viewportData.getItem(keyedItem.key) != null) {
+        viewportData.update(keyedItem.key, keyedItem);
+      } else {
+        viewportData.append(keyedItem);
+      }
+    });
+  };
+}
+
+/**
+ * Maps a Row to a key / value object. Keys are mapped ver batim from column
+ * names.
+ * @param row Row to map to an item.
+ * @param columns Columns to map.
+ * @returns A key / value object for the row.
+ */
+export function defaultRowDeserializer<T>(
+  row: ViewportRow,
+  columns: Column[]
+): T {
+  return columns.reduce((result, col) => {
+    // eslint-disable-next-line no-param-reassign
+    result[col.name as keyof T] = row.get(col);
+    return result;
+  }, {} as T);
+}
+
+/**
+ * For windowing to work, the underlying list needs to maintain a KeyedItem for
+ * each row in the backing table (even if these rows haven't been loaded yet).
+ * This is needed internally by react-spectrum so it can calculate the content
+ * area size. This generator can create a range of empty `KeyedItem` objects.
+ * @param count The number of items to generate
+ */
+export function* generateEmptyKeyedItems<T>(
+  count: number
+): Generator<KeyedItem<T>, void, unknown> {
+  // eslint-disable-next-line no-plusplus
+  for (let i = 0; i < count; ++i) {
+    yield { key: String(i) };
+  }
+}
+
+/**
+ * Check a Table to see if it is closed before checking its size property. This
+ * is important because calling Table.size on a closed table throws an error. If
+ * the table is null or closed, return zero. Otherwise, return the current size.
+ * @param table The table to check for its size.
+ * @returns The size of the table or zero if the table is null or closed.
+ */
+export function getSize(table?: Table | TreeTable | null): number {
+  return table == null || isClosed(table) ? 0 : table.size;
+}
+
+/**
+ * Check if a given table is closed. Tree tables don't have an `isClosed` prop,
+ * so will always return false.
+ * @param table The table to check if it is closed.
+ */
+export function isClosed(table: Table | TreeTable): boolean {
+  if ('isClosed' in table) {
```

Comment:
I'd add a following ticket in deephaven-core to add the `isClosed` property to `TreeTable`. Especially since it already has a `closed` variable: https://github.com/deephaven/deephaven-core/blob/226a428be5e746a2f1fbf07264a742769631c4c2/web/client-api/src/main/java/io/deephaven/web/client/api/tree/JsTreeTable.java#L358
... and that `closed` variable never gets set. So something seems amiss there. This check is good though.

Reply:
I created https://github.com/deephaven/deephaven-core/issues/3714 to follow up on this.